goSrc/src/vmCaller/rpc/eth_service.go

```diff
@@ -0,0 +1,1034 @@
+func (srv *EthService) Web3Sha3(req *web3.Web3Sha3Params) (*web3.Web3Sha3Result, error) {
+	data, err := x.DecodeToBytes(req.Data)
+	if err != nil {
+		return nil, err
+	}
+
+	return &web3.Web3Sha3Result{
+		HashedData: x.EncodeBytes(crypto.Keccak256(data)),
+	}, nil
+}
+
+// NetListening returns true if the peer is running
+func (srv *EthService) NetListening() (*web3.NetListeningResult, error) {
+	return &web3.NetListeningResult{
+		IsNetListening: srv.nodeView.NodeInfo().GetListenAddress() != "",
+	}, nil
+}
+
+// NetPeerCount returns the number of connected peers
+func (srv *EthService) NetPeerCount() (*web3.NetPeerCountResult, error) {
+	return &web3.NetPeerCountResult{
+		NumConnectedPeers: x.EncodeNumber(uint64(srv.nodeView.Peers().Size())),
+	}, nil
+}
+
+// NetVersion returns the hex encoding of the network id,
+// this is typically a small int (where 1 == Ethereum mainnet)
+func (srv *EthService) NetVersion() (*web3.NetVersionResult, error) {
+	return &web3.NetVersionResult{
+		ChainID: x.EncodeNumber(uint64(chainID)),
+	}, nil
+}
+
+// EthProtocolVersion returns the version of tendermint
+func (srv *EthService) EthProtocolVersion() (*web3.EthProtocolVersionResult, error) {
+	return &web3.EthProtocolVersionResult{
+		ProtocolVersion: srv.nodeView.NodeInfo().Version,
+	}, nil
+}
+
+// EthChainId returns the chainID
+func (srv *EthService) EthChainId() (*web3.EthChainIdResult, error) {
+	return &web3.EthChainIdResult{
+		ChainId: srv.blockchain.ChainID(),
+	}, nil
+}
+
+// EthBlockNumber returns the latest height
+func (srv *EthService) EthBlockNumber() (*web3.EthBlockNumberResult, error) {
+	return &web3.EthBlockNumberResult{
+		BlockNumber: x.EncodeNumber(srv.blockchain.LastBlockHeight()),
+	}, nil
+}
+
+// EthCall executes a new message call immediately without creating a transaction
+func (srv *EthService) EthCall(req *web3.EthCallParams) (*web3.EthCallResult, error) {
+	var to crypto.Address
+	var from string
+	var err error
+
+	if addr := req.Transaction.To; addr != "" {
+		to, err = x.DecodeToAddress(addr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if addr := req.Transaction.From; addr != "" {
+		from = addr
+	}
+
+	data, err := x.DecodeToBytes(req.Transaction.Data)
+	if err != nil {
+		return nil, err
+	}
+	txe, err := myExecution.CallSim(srv.accounts, srv.blockchain, from, to, data, srv.logger)
+	if err != nil {
+		return nil, err
+	} else if txe.Exception != nil {
+		return nil, txe.Exception.AsError()
+	}
+
+	var result string
+	if r := txe.GetResult(); r != nil {
+		result = x.EncodeBytes(r.GetReturn())
+	}
+
+	return &web3.EthCallResult{
+		ReturnValue: result,
+	}, nil
+}
+
+// EthGetBalance returns an accounts balance, or an error if it does not exist
+func (srv *EthService) EthGetBalance(req *web3.EthGetBalanceParams) (*web3.EthGetBalanceResult, error) {
+	addr, err := x.DecodeToAddress(req.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: read account state at height
```

Comment:
what with the TODO? can't it be done in the PR? If not ask please iroha core team if the TODO can stay in the new code?